chore(build): Add LOTS of warnings

anthonybilinski opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/6551)
<!-- Reviewable:end -->

robinlinden commented on CMakeLists.txt:
```diff
@@ -104,6 +104,7 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wstrict-aliasing")
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wold-style-cast")
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wsign-compare")
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wunknown-pragmas")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wswitch")
```
I was about to write that this changes nothing as `-Wswitch` is part of the **basic** `-Wall` warnings, but I guess qTox doesn't have those enabled.. :(

anthonybilinski replied:
Good point :grimacing: 

Enabled a ton stolen from you and beat qTox into compliance. Updated with a huge diff.

Each refactor commit should be non-functional (except for trivial things like using float precision instead of double precision in UI). Functional commits are marked as "fix".

anthonybilinski committed:
chore(build): Enable lots of compile-time warnings

Taken directly from https://github.com/evilcorpltd/n_e_s/tree/9777aa619db619c1488fc016b0912a0efb2cc303/warnings

* Remove Wuseless-cast, because Qt's MOC-generated code hits it.
* Remove Wduplicated-branches, Wduplicated-branches, Wformat-truncation
because they aren't supported on our oldest CI job.

anthonybilinski committed:
chore(build): Remove redundant warnings

anthonybilinski committed:
chore(build): Migrate existing warning handling to new warnings project

anthonybilinski committed:
refactor: Comply with Wshadow

Avoid shadowing variables:
* Rename variables to something better if possible
* If not, postfix shadowing arguments with _. Favour leaving member
  variables without postfixes.
* Rename variables prefixed with _ to avoid library function collisions
* Avoid double underscore anywhere in names
* Make definition and declaration argument names match where seen
* Favour using class variable over argument variable, where equivalent
* Remove explicit this-> where equivalent

anthonybilinski committed:
fix(ui): Use GenericChatItemWidget construction compact argument

anthonybilinski committed:
fix(widget): Update widget's core on core update

We no longer support changing cores, but might as well not leave it in a
clearly broken state.

anthonybilinski committed:
refactor: Comply with Wreorder

Don't reorder constructor init from class claration order.

anthonybilinski committed:
refactor: Comply with Wunused-but-set-variable

Remove unused variables.

anthonybilinski committed:
refactor: Comply with Wdeprecated-copy

Remove implicitly declared copy constructors

anthonybilinski committed:
fix(ui): Pass NetCamView's parent pointer to base QWidget

anthonybilinski committed:
refactor: Comply with Wunused-parameter

Explicitly ignore unused parameters.

Also replace Q_UNUSED with std::ignore for consistency and to favour std
over Qt when equivalent.

anthonybilinski committed:
refactor: Comply with Wmissing-field-initializers

Add explicit missing initializers.

anthonybilinski committed:
refactor: Comply with Wdouble-promotion

Remove implicit float to double conversions.

anthonybilinski committed:
refactor: Comply with Wmissing-declarations

* Move free functions to the anonymous namespace
* Additionally move static free functions to the anonymous namespace
* Move functions that must be accessed externally to static class functions

anthonybilinski committed:
refactor: Comply with Wundef

Don't try to read undefined macros.

anthonybilinski committed:
refactor: Remove some useless casts

Can't actually enable Wuseless-cast though because Qt's MOC-generated code is
full of them.

anthonybilinski committed:
refactor: Comply with Wpedantic

* Remove semicolons after calling a macro
* Remove semicolons at the end of class declarations
* Remove semicolons at the end of namespaces

anthonybilinski committed:
refactor: Comply with Wmaybe-uninitialized

anthonybilinski committed:
refactor: Comply with Wnull-dereference

* Add check before dereferencing return from qobject_cast

anthonybilinski committed:
refactor: Comply with Wformat
